### DISCUSSION TITLE
Snippets

### DISCUSSION THREAD

**Discussion by raylemon** [abougouffa/minemacs#101] [Q&A] [answered]
Bonjour,
Existe-il un moyen pour créer des snippets  ?

Merci !

**Comment by abougouffa** [comment, marked as answer]
J'utilise [`tempel` dans MinEmacs](https://github.com/abougouffa/minemacs/blob/59b486445ef1d2cdd1e9f57de222d45b8b8582e2/modules/me-editor.el#L11-L25). [Le format des snippets Tempel](https://github.com/minad/tempel#template-file-format) est très simple et MinEmacs [inclut aussi `tempel-collection`](https://github.com/abougouffa/minemacs/blob/59b486445ef1d2cdd1e9f57de222d45b8b8582e2/modules/me-editor.el#L27-L30) qui rassemble quelques snippets prédéfinies. J'utilise les snippets très occasionnellement, donc je n'ai pas de collection personnelle prête à utiliser.

Par défaut, pour compléter les snippets, tu (je pense qu'on peut se tutoyer) dois écrire `<` (facilement accessible sur un clavier Azerty), use liste de complétion doit apparaitre avec les abréviations des snippets disponibles pour le mode actif.

Il y a aussi un sous-module obsolète à base de `yasnippet` que tu peux activer en ajoutant `obsolete/me-yasnippet` à la liste `minemacs-modules`, néanmois, cette config est obsolète et elle n'est pas testée.

**Comment by raylemon** [comment]
Super, merci. J’aimerais ajouter des snippets, mais je ne suis pas très habitué au lisp ^^.  Si je comprends bien, je dois créer un fichier eld que je dois mettre dans le dossier `templates` dans .minemacs.d, c’est ça ?

**Comment by abougouffa** [comment]
Avec le dernier commit 59d8b1fd79e32ef21fb0d497c7245894ca88ef78, tu dois être capable de faire ça. Met les fichiers `.eld` dans `.minemacs.d/templates/tempel/` et ils seront reconnus par `tempel`.

**Comment by raylemon** [comment]
Super, ça fonctionne !! Merci beaucoup !
J’adore ta config emacs, je la trouve plus rapide et plus simple que l’usine à gaz DOOM.
Existe-t-il, comme dans DOOM, une macro ou une variable permettant de tester l’os ? DOOM possède IS-LINUX, IS-MAC, IS-WINDOWS.

Ça permettrait de créer des configurations alternatives selon l’OS. Dans mon cas, je pourrais désactiver eglot sous Windows et utiliser le package languagetool (plus configurer deux ou trois trucs qui posent problème sous uniquement sous Windows…) et garder la config de base sous Linux (et WSL)

**Reply by abougouffa** [reply]
Merci, heureux de savoir que tu l'as trouvé utile!

Yep, il y a des constantes équivalentes à celles de Doom définies dans `me-vars`:

https://github.com/abougouffa/minemacs/blob/59d8b1fd79e32ef21fb0d497c7245894ca88ef78/core/me-vars.el#L104-L107

**Comment by raylemon** [comment]
Super, merci. Ainsi je peux désactiver eglot sur Windows. Autre question, tu utilises juste eglot ou aussi eglot-ltex ? C’est pour savoir sur quel repo je dois remonter le bug…

**Reply by raylemon** [reply]
Comment je peux utiliser les constantes ? Si je fais (if (os/win) …), il me dit 

> Symbol’s function definition is void: os/win

**Reply by abougouffa** [reply]
Non, ce bout `(if (os/win) ...)` suppose que `os/win` est une **fonction**, alors que c'est une constante/variable, donc il faut `(if os/win ...)` sans les parenthèses. En gros, les parenthèses en Lisp c'est pour évoquer une fonction,

Donc sur Lisp:

**`(`** `nom-de-fonction` `arg1` `arg2` **`)`** 

est équivalent à une appelle de fonction sur C, Python, ... de type:

`nom-de-fonction` **`(`** `arg1` `,` `arg2` **`)`**

C'est juste la position de la parenthèse gauche qui est différente en fait!

**Reply by raylemon** [reply]
Sans les parenthèses ? Je l’ai fait également, mais ça m’affichait le même problème…
J’ai écrit ceci dans mon config.el

`(if os/win
  (setopt python-shell-interpreter "python"))`
  
Cependant, je viens de remarque que parinfer a ajouté un peu trop de parenthèses… j’ai dû en oublier une quelque part…
